### PR TITLE
Support scrub to zoom for alternate charts

### LINKED ISSUE #4857: Feat: Scrub to Zoom for alternate charts
Add support for scrub to zoom for alternate charts

https://www.notion.so/rilldata/Enable-scrub-to-filter-on-measure-line-charts-5a662647d43e4e23937a709145bbdd9c
https://www.notion.so/rilldata/UXQA-Alt-Viz-fb97d9eb1c5c4fb1bfb3328e83008172

Notes
- “scrubbing” or clicking and dragging across the chart
- MeasureChart - line (d3)
- TDDAlternateChart - others (vega-lite)
- MeasureScrub.svelte

Interaction
- https://vega.github.io/vega-lite/docs/selection.html#current-limitations
- Scrub demo: https://observablehq.com/@jonfroehlich/intro-to-interaction-in-vega-lite
- Click-drag and scroll to pan and zoom. Hold the shift key and click-drag for brushing. https://observablehq.com/@vega/vega-lite-composing-selections?collection=@vega/vega-lite-api
- Interval X: https://observablehq.com/@john-guerra/vega-lite-selection-methods

Related
- https://github.com/rilldata/rill/pull/2877
- https://github.com/rilldata/rill/pull/2967
- https://github.com/rilldata/rill/issues/4857
- https://github.com/rilldata/rill/pull/2131

### PR DESCRIPTION
**Pull request by lovincyrus**
Closes #4857

- [x] enable interval selection
- [x] how to capture the interval values?
- [x] bind the intervals to `$dashboardStore?.selectedScrubRange?.start and $dashboardStore?.selectedScrubRange?.end`
- [x] `zoomScrub` based on the scrub range then `setSelectedTimeRange`
- [x] enable brush for stacked area and stacked bar
- [x] clear interval on esc
- [x] only set scrub range on drag end 
- [x] explore [full height hover on bar chart](https://vega.github.io/editor/#/examples/vega-lite/interactive_histogram_full_height_hover)
- [x] access brushend
- [x] check isScrubbing for alt charts (blocked by the ability to access brushend)
- [x] fix band size
- [x] check bar band relative size configs in vega and vega lite specs
- [x] update bars style that are in the selected interval
- [x] disable tooltip for the scrubbed interval range (configurable via global bar tooltip config)
- [x] check vega chart theme
- [x] unable to scrub bar chart when comparison is enabled
- [x] esc to clear the selection interval
- [x] revert full height hover support, can land scrub to zoom first
- [x] fix updateVegaOnTableHover unrecognized "hover_tuple_fields"
- [x] clear the interval even when switching chart type
- [x] configure other chart types' color based on brush param
- [x] to prune diff changed
- [x] fix WARN Config.customFormatTypes is not true, thus custom format type and format for channel y are dropped.
- [x] remove tooltip when isScrubbing https://github.com/rilldata/rill/pull/5483#issuecomment-2330238319
- [x] fix precision of scrub range https://github.com/rilldata/rill/pull/5483#issuecomment-2330244184
- [x] disable tooltip when scrubbing
- [x] fix https://github.com/rilldata/rill/pull/5483#issuecomment-2335121414

Refs https://github.com/vega/vega-lite/issues/3338, https://github.com/vega/vega-lite/issues/5341, https://github.com/vega/vega-lite/releases/tag/v5.0.0

Update 1: Vega-Lite does not offer direct access to the signals API. To handle events, we would need to compile the Vega-Lite spec into Vega and then add custom signals. To simplify event handling, I’ve opted to use JavaScript to dispatch and manage changes. See: https://github.com/rilldata/rill/pull/5483/commits/59b42ea510c317b3af0138d54ff200d3d5aa3b00

Update 2: Compile vega lite spec → vega spec and add vega signals to configure brushend so we can configure the `isScrubbing` state. https://github.com/rilldata/rill/pull/5483/commits/74471034d8a8afdb11f61ed3011aff8ff71a77de

Update 3: Fun fact: Vega spec doesn't support relative bar width band size.

> In Vega-Lite, bar widths can be adjusted using the bandSize property for banded scales, which effectively controls the width of bars in a banded scale. However, Vega itself does not provide a direct bandSize equivalent. Instead, Vega relies on a combination of encoding channels and layout configurations to control bar widths.

> In Vega, you can use scale with domain and range to control the width of bars, but you might need to calculate and set these values dynamically if you need relative sizing. You can also use mark properties and layout settings to manage the width and spacing of bars.

> So, while Vega-Lite provides a convenient way to handle relative bar widths with bandSize, Vega requires a more manual approach to achieve similar results.

Update 4:

https://github.com/user-attachments/assets/0bd8baa1-5c0a-4be6-b316-5ca1a123493a

Update 5: Keeping vega lite → vega spec changes to support more advanced interactions down the road

### CODE REVIEW

**Comment by djbarnwal** [comment]
Did an initial UXQA pass

1. Default bars look a bit lighter than they are on main
image

2. The color of the bars when comparing with time is now different
image

3. The zoom range and the colored bar range is not the same (8AM should be inclusive and 3PM should be exclusive)
image

4. Can't see tooltip on the bars which are inside the scrub range.

**Comment by lovincyrus** [comment]
> Did an initial UXQA pass
> 
> 1. Default bars look a bit lighter than they are on main
>    image
> 2. The color of the bars when comparing with time is now different
>    image
> 3. The zoom range and the colored bar range is not the same (8AM should be inclusive and 3PM should be exclusive)
>    image
> 4. Can't see tooltip on the bars which are inside the scrub range.

Fixed 1 and 2.

3 - Looks like the entire scrub range is being taken into consideration rather than just the highlighted bars. This provides more granular controls since we can even start scrubbing from the whitespace.

https://github.com/user-attachments/assets/362feaa5-9407-4ee7-999d-1b91299d2348

The current measure chart's behavior is snapping to a fixed point. Do you think we should filter out the highlighted values within the scrub range or keep the current behavior? @djbarnwal 

4 - This appears to be a known limitation of vega library when hovering over the brush selection layer. https://github.com/vega/vega-lite/issues/6003 We can explore workarounds in the follow up if needed.

**Comment by djbarnwal** [comment]
Also, we should disable tooltips when scrubbing, otherwise we see a tooltip as we scrub. Also at the end of scrub the tooltip for the last bar still remains

**Comment by lovincyrus** [comment]
> Also, we should disable tooltips when scrubbing, otherwise we see a tooltip as we scrub.

Absolutely! I tried to disable the tooltip during scrubbing, but unfortunately, it didn't work. Here's the past commit: https://github.com/rilldata/rill/pull/5483/commits/20ec4c221d5f3648e9400a9df819943cd0f2997d

https://github.com/vega/vega-lite/issues/8750
https://vega.github.io/vega-lite/docs/tooltip.html#disable-tooltips
https://github.com/vega/altair/issues/3196

```
    tooltip: [
      {
        condition: {
          test: "!length(data('brush_store'))",
          field: timeField.tooltipName ? timeField.tooltipName : timeField.name,
          type: "temporal",
          title: "Time",
          format: "%b %d, %Y %H:%M",
        },
      },
      {
        condition: {
          test: "!length(data('brush_store'))",
          title: quantitativeField.label,
          field: quantitativeField.name,
          type: "quantitative",
          formatType: quantitativeField.formatterFunction || "number",
        },
      },
    ],
```

```
tooltip: {
      condition: {
        param: "brush",
        value: [
          {
            field: timeField.tooltipName
              ? timeField.tooltipName
              : timeField.name,
            type: "temporal",
            title: "Time",
            format: "%b %d, %Y %H:%M",
          },
          {
            title: quantitativeField.label,
            field: quantitativeField.name,
            type: "quantitative",
            formatType: quantitativeField.formatterFunction || "number",
          },
        ],
      },
      value: null,
    },
```

```
    tooltip: isScrubbing
      ? undefined
      : [
          {
            field: timeField.tooltipName
              ? timeField.tooltipName
              : timeField.name,
            type: "temporal",
            title: "Time",
            format: "%b %d, %Y %H:%M",
          },
          {
            title: quantitativeField.label,
            field: quantitativeField.name,
            type: "quantitative",
            formatType: quantitativeField.formatterFunction || "number",
          },
        ],
```

Setting tooltip to false in vega config also didn't work.
```
  mark: {
    tooltip: isScrubbing ? false : undefined,
  },
```

If we want, we could technically hide the tooltip visually as a workaround as well
```
  :global(#rill-vg-tooltip) {
    @apply hidden;
  }
```

**Comment by lovincyrus** [comment]
> For alternate charts, it's a bit tricky. For our current line charts, the points are plotted in the center so we can snap on the edges of the bucketed period. For vega charts, the bars are placed just below the axis labels.

Indeed! The source of truth of the vega chart is the interval.

CleanShot 2024-09-04 at 15 10 10@2x

We should ensure that the scrubbed range is accurately reflected in the highlighted cells.﻿

CleanShot 2024-09-04 at 14 26 11@2x

**Comment by lovincyrus** [comment]
> > Also, we should disable tooltips when scrubbing, otherwise we see a tooltip as we scrub.
> 
> Absolutely! I tried to disable the tooltip during scrubbing, but unfortunately, it didn't work. Here's the past commit: [20ec4c2](https://github.com/rilldata/rill/commit/20ec4c221d5f3648e9400a9df819943cd0f2997d)
> 
> [vega/vega-lite#8750](https://github.com/vega/vega-lite/issues/8750) https://vega.github.io/vega-lite/docs/tooltip.html#disable-tooltips [vega/altair#3196](https://github.com/vega/altair/issues/3196)
> 
> ```
>     tooltip: [
>       {
>         condition: {
>           test: "!length(data('brush_store'))",
>           field: timeField.tooltipName ? timeField.tooltipName : timeField.name,
>           type: "temporal",
>           title: "Time",
>           format: "%b %d, %Y %H:%M",
>         },
>       },
>       {
>         condition: {
>           test: "!length(data('brush_store'))",
>           title: quantitativeField.label,
>           field: quantitativeField.name,
>           type: "quantitative",
>           formatType: quantitativeField.formatterFunction || "number",
>         },
>       },
>     ],
> ```
> 
> ```
> tooltip: {
>       condition: {
>         param: "brush",
>         value: [
>           {
>             field: timeField.tooltipName
>               ? timeField.tooltipName
>               : timeField.name,
>             type: "temporal",
>             title: "Time",
>             format: "%b %d, %Y %H:%M",
>           },
>           {
>             title: quantitativeField.label,
>             field: quantitativeField.name,
>             type: "quantitative",
>             formatType: quantitativeField.formatterFunction || "number",
>           },
>         ],
>       },
>       value: null,
>     },
> ```
> 
> ```
>     tooltip: isScrubbing
>       ? undefined
>       : [
>           {
>             field: timeField.tooltipName
>               ? timeField.tooltipName
>               : timeField.name,
>             type: "temporal",
>             title: "Time",
>             format: "%b %d, %Y %H:%M",
>           },
>           {
>             title: quantitativeField.label,
>             field: quantitativeField.name,
>             type: "quantitative",
>             formatType: quantitativeField.formatterFunction || "number",
>           },
>         ],
> ```
> 
> Setting tooltip to false in vega config also didn't work.
> 
> ```
>   mark: {
>     tooltip: isScrubbing ? false : undefined,
>   },
> ```
> 
> If we want, we could programmatically hide the tooltip visually as a workaround as well
> 
> ```
>   :global(#rill-vg-tooltip) {
>     @apply hidden;
>   }
> ```

Okay, fixed this with https://github.com/rilldata/rill/pull/5483/commits/a22c942382b1fb704aa5799682a7976807884cae by configuring the tooltip handler directly.

https://github.com/user-attachments/assets/c944065a-f0d9-4f8e-a279-439030b1214f

**Comment by lovincyrus** [comment]
Addressed the feedback, @djbarnwal 

I think this is a good baseline. I have created some follow up tickets to keep track of the remaining issues so we can land them incrementally.

https://github.com/rilldata/rill/issues/5633, https://github.com/rilldata/rill/issues/5634

**Comment by djbarnwal** [comment]
Have mentioned a few small but critical bugs in this loom video 

https://www.loom.com/share/62bc766c466847608ced7c1eee604860?sid=dc47bbad-77a3-40d2-bfd2-49214aabd1d1

File: web-common/src/features/charts/render/vega-config.ts [hunk 1]
```diff
@@ -16,9 +16,6 @@ export const getRillTheme: () => Config = () => ({
   autosize: {
     type: "fit-x",
   },
-  mark: {
-    tooltip: true,
-  },
```
Comment (djbarnwal) [inline web-common/src/features/charts/render/vega-config.ts:21]: What's the idea behind removing this? We also use the same config for our custom dashboard charts so this removes the default tooltips from those charts.

File: web-common/src/features/dashboards/time-dimension-details/charts/TDDAlternateChart.svelte [hunk 2]
```diff
@@ -34,6 +39,7 @@
   export let isTimeComparison: boolean;
 
   let viewVL: View;
+  let vegaSpec: any;
```
Comment (djbarnwal) [inline web-common/src/features/dashboards/time-dimension-details/charts/TDDAlternateChart.svelte:42]: Can we add the type for Vega Spec here? (should be `VisualizationSpec` or `VgSpec`)

File: web-common/src/features/dashboards/time-dimension-details/charts/utils.ts [hunk 3]
```diff
@@ -132,11 +137,46 @@ export function updateVegaOnTableHover(
...
+      if (param.name === "brush") {
+        return true;
+      }
+    }
+  }
+  return false;
+}
+
+// Check if vega lite spec has brush params
+// If so, we need toompile vega lite spec to vega spec
```
Comment (djbarnwal) [inline web-common/src/features/dashboards/time-dimension-details/charts/utils.ts:163]: nit: typo

File: web-common/src/features/charts/render/vega-config.ts (hunk 1 above)
Comment (lovincyrus) [reply web-common/src/features/charts/render/vega-config.ts:21]: > What's the idea behind removing this?

Keeping this in the global config will trigger an empty tooltip on the brush interval.

```
  mark: {
    tooltip: true,
  },
```


https://github.com/user-attachments/assets/cdda169d-bd64-44a7-bb9a-b55070425edf

Comment (lovincyrus) [reply web-common/src/features/charts/render/vega-config.ts:21]: If the global tooltip config is absolutely needed, I can enable the tooltip flag for customDashboard.

Comment (lovincyrus) [reply web-common/src/features/charts/render/vega-config.ts]: https://github.com/rilldata/rill/pull/5483/commits/86805de59ca62a3ccfeb317ad52e8eac113dec4b

**Comment by lovincyrus** [comment]
> Have mentioned a few small but critical bugs in this loom video
> 
> https://www.loom.com/share/62bc766c466847608ced7c1eee604860?sid=dc47bbad-77a3-40d2-bfd2-49214aabd1d1

Nice catch on the timezone inconsistencies! I'll review if any of the code paths affected these behaviors.

**Comment by lovincyrus** [comment]
> > Have mentioned a few small but critical bugs in this loom video
> > https://www.loom.com/share/62bc766c466847608ced7c1eee604860?sid=dc47bbad-77a3-40d2-bfd2-49214aabd1d1
> 
> Nice catch on the timezone inconsistencies! I'll review if any of the code paths affected these behaviors.

Looking back at the scrub behaviors of the granular details of time range, we weren't able to scrub the chart at all when the time grain is `TIME_GRAIN_HOUR`

CleanShot 2024-09-09 at 09 36 42@2x

https://github.com/user-attachments/assets/cd946e96-ef9d-4091-ae29-c14fbda5c6f9

**Comment by lovincyrus** [comment]
Fixed timezone inconsistency in the alt charts

https://github.com/user-attachments/assets/36c5e587-ccae-427b-b549-6574be32044f

CleanShot 2024-09-09 at 17 29 54@2x

File: web-common/src/features/dashboards/time-dimension-details/charts/TDDAlternateChart.svelte [hunk 4]
```diff
@@ -117,13 +211,26 @@
...
-{#if sanitizedVegaSpec && data}
+{#if hasBrushParam(sanitizedVegaLiteSpec) && data}
+  <VegaRenderer
+    bind:view={viewVL}
+    data={{ table: data }}
+    spec={vegaSpec}
+    {signalListeners}
+    {expressionFunctions}
+    {tooltipFormatter}
+    canvasDashboard={$customDashboards}
```
Comment (djbarnwal) [inline web-common/src/features/dashboards/time-dimension-details/charts/TDDAlternateChart.svelte:222]: This should be false and not rely on the feature flag.

File: web-common/src/features/dashboards/time-dimension-details/charts/TDDAlternateChart.svelte [hunk 5]
```diff
@@ -117,13 +211,26 @@
...
+  <!-- JIC we add a new chart type without brush param -->
   <VegaLiteRenderer
     bind:viewVL
     data={{ table: data }}
-    spec={sanitizedVegaSpec}
+    spec={sanitizedVegaLiteSpec}
     {signalListeners}
     {expressionFunctions}
     {tooltipFormatter}
+    canvasDashboard={$customDashboards}
```
Comment (djbarnwal) [inline web-common/src/features/dashboards/time-dimension-details/charts/TDDAlternateChart.svelte:234]: This should be false and not rely on the feature flag.

File: web-common/src/features/canvas-components/render/VegaRenderer.svelte [hunk 6]
```diff
@@ -0,0 +1,153 @@
...
+  import { onDestroy } from "svelte";
+  import { getRillTheme } from "./vega-config";
+
+  export let data: Record<string, unknown> = {};
+  export let spec: VisualizationSpec;
+  export let signalListeners: SignalListeners = {};
+  export let expressionFunctions: ExpressionFunction = {};
+  export let error: string | null = null;
+  export let canvasDashboard = false;
+  export let chartView = false;
```
Comment (ericpgreen2) [inline web-common/src/features/canvas-components/render/VegaRenderer.svelte:23]: It seems a little off for a "Vega[Lite]Renderer" to be aware of canvas dashboards. I suppose I'd expect that  a "Vega[Lite]Renderer" would solely render Vega visualizations, and any Rill application logic (e.g. references to canvases or the runtime) would be handled outside of it.

(Though I noticed that the `VegaLiteRenderer` also includes a `canvasDashboard` prop, so I do agree with keeping the two Vega renderers symmetrical.)

Comment (lovincyrus) [reply web-common/src/features/canvas-components/render/VegaRenderer.svelte:23]: https://rilldata.slack.com/archives/C04AGHL77PS/p1726514541389759?thread_ts=1726509359.103919&cid=C04AGHL77PS

**Review by djbarnwal — APPROVED** [review]
Thanks for the PR!


A bug I noticed while playing around was that for some scrubbbed ranges, the table incorrectly shows no data. We can address this in a follow up burndown list.

image

**Pull request merged**